src/analysis/processing/qgsalgorithmb3dmtogltf.cpp

```diff
@@ -0,0 +1,139 @@
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsalgorithmb3dmtogltf.h"
+#include "qgscesiumutils.h"
+#include "qgsgltfutils.h"
+
+#define TINYGLTF_IMPLEMENTATION
+#define TINYGLTF_NO_STB_IMAGE         // we use QImage-based reading of images
+#define TINYGLTF_NO_STB_IMAGE_WRITE   // we don't need writing of images
+
+#include "tiny_gltf.h"
+#include <fstream>
+
+///@cond PRIVATE
+
+QString QgsB3DMToGltfAlgorithm::name() const
+{
+  return QStringLiteral( "b3dmtogltf" );
+}
+
+QString QgsB3DMToGltfAlgorithm::displayName() const
+{
+  return QObject::tr( "Convert B3DM to GLTF" );
+}
+
+QStringList QgsB3DMToGltfAlgorithm::tags() const
+{
+  return QObject::tr( "3d,tiles,cesium" ).split( ',' );
+}
+
+QString QgsB3DMToGltfAlgorithm::group() const
+{
+  return QObject::tr( "3D Tiles" );
+}
+
+QString QgsB3DMToGltfAlgorithm::groupId() const
+{
+  return QStringLiteral( "3dtiles" );
+}
+
+QString QgsB3DMToGltfAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Converts files from the legacy B3DM format to GLTF." );
+}
+
+QgsB3DMToGltfAlgorithm *QgsB3DMToGltfAlgorithm::createInstance() const
+{
+  return new QgsB3DMToGltfAlgorithm();
+}
+
+void QgsB3DMToGltfAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFile( QStringLiteral( "INPUT" ), QObject::tr( "Input B3DM" ), QgsProcessingParameterFile::File,
+                QStringLiteral( "b3dm" ), QVariant(), false, QStringLiteral( "B3DM (*.b3dm *.B3DM)" ) ) );
+
+  addParameter( new QgsProcessingParameterFileDestination( QStringLiteral( "OUTPUT" ), QObject::tr( "Output GLTF file" ), QStringLiteral( "GLTF (*.gltf *.GLTF)" ) ) );
+}
+
+QVariantMap QgsB3DMToGltfAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  const QString path = parameterAsFile( parameters, QStringLiteral( "INPUT" ), context );
+  const QString outputPath = parameterAsFile( parameters, QStringLiteral( "OUTPUT" ), context );
+
+  if ( !QFile::exists( path ) )
+    throw QgsProcessingException( QObject::tr( "Could not load source file %1." ).arg( path ) );
+
+  QFile f( path );
+  QByteArray b3dmContent;
+  if ( f.open( QIODevice::ReadOnly ) )
+  {
+    b3dmContent = f.readAll();
+  }
+  else
+  {
+    throw QgsProcessingException( QObject::tr( "Could not load source file %1." ).arg( path ) );
+  }
+
+  const QByteArray gltfContent = QgsCesiumUtils::extractGltfFromB3dm( b3dmContent );
+
+  // load gltf and then rewrite -- this allows us to both validate the B3DM GLTF content, and
+  // also gives the opportunity to "upgrade" B3DM specific options (like CESIUM_RTC) to standard
+  // GLTF extensions
```

Comment:
This would be added once #54291 is merged, right?

By the way, with CESIUM_RTC from b3dm, there are still actually two options to store it in GLTF - either as a CESIUM_RTC extension, or (like google does) just simple translation of the root node.

Reply:
> This would be added once https://github.com/qgis/QGIS/pull/54291 is merged, right?

Correct

> By the way, with CESIUM_RTC from b3dm, there are still actually two options to store it in GLTF - either as a CESIUM_RTC extension, or (like google does) just simple translation of the root node.

I was going to go for the upgrade to the extension approach -- that seems like the most direct equivalent.